File: boj/#14716.cpp
#include<iostream>
#include<queue>
#include<algorithm>

using namespace std;

#define MAX 251
#define DIR 8

int N, M;
int way[8][2] = { {1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {1, -1}, {-1, 1} };
bool board[MAX][MAX];

void bfs(pair<int, int> start_point) {
	queue<pair<int, int>> q;

	board[start_point.first][start_point.second] = 0;
	q.push(start_point);

	while (!q.empty()) {
		int curi = q.front().first;
		int curj = q.front().second;
		q.pop();
		
		for (int d = 0; d < DIR; d++) {
			int nexti = curi + way[d][0];
			int nextj = curj + way[d][1];

			// 범위 check
			if (nexti < 0 || nextj < 0 || nexti >= N || nextj >= M)
				continue;

			if (board[nexti][nextj] == 1) {
				board[nexti][nextj] = 0;
				q.push({ nexti, nextj });
			}
		}
	}
	return;
}

int main() {
	
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	cin >> N >> M;

	int ans = 0;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> board[i][j];
		}
	}
	
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if(board[i][j] == 1){
				bfs({ i, j });

				ans++;
			}
		}
	}

	cout << ans << "\n";
	
	return 0;
}

File: boj/#14716.py
from collections import deque

way = [[0, 1], [1, 0], [-1, 0], [0, -1], [1, 1], [-1, -1], [-1, 1], [1, -1]]

def bfs(point):
    q = deque()

    q.append(point)
    board[point[0]][point[1]] = 0
    
    while len(q) > 0:
        curi, curj = q.popleft()

        for i, j in way:
            nexti, nextj = curi + i, curj + j

            if nexti < 0 or nextj < 0 or nexti >= N or nextj >= M:
                continue

            if board[nexti][nextj] == 1:
                board[nexti][nextj] = 0
                q.append([nexti, nextj])

        
   

N, M = list(map(int, input().split()))

board = [list(map(int, input().split())) for _ in range(N)]

ans = 0

for i in range(0, N):
    for j in range(0, M):
        if board[i][j] == 1:
            bfs([i, j])
            ans += 1

print(ans)